client/client.go

```diff
@@ -136,77 +238,107 @@ func decodeHeader(header []byte) (code []byte, pt uint32, length int) {
 	return
 }
 
-func (client *Client) reconnect(err error) error {
-	if client.conn != nil {
-		return nil
+func (client *Client) grabReconnectState() (success bool) {
+	client.reconnectLock.Lock()
```

Comment:
I am not clear why you would use a mutex here instead of CAS. You are doing a compare and swap in this operation, why not push it down to the memory controller by using `sync/atomic.CompareAndSwapUnit32`? i.e.

```
func (client *Client) lockReconnect() bool {
    return atomic.CompareAndSwapUint32(&client.reconnecting, 0, 1)
}
```

This has the same behavior and is less expensive (and less code).